File: binary-search.py
# Code walkthrough 
# Soliving in python 

##recursive binary search 



def binarySearch(array, target):
return binarySearchHelper(array, target, 0, array - 1)


def binarySearchHelper(array, target, left, right):
# base case is our right pointer greater than our left pointer
while left <= right:
    middle = (left + right) // 2
    # potental match number in arrray located in middle 
    potentialMatch = array[middle]
    if target === potentialMatch:
        # return our middle index
        return middle
    elif target < potentialMatch:
        right = middle -1
    else:
        left = middle + 1
return -1
        




# [141, 1, 17, -7, -17, -27, 18, 541, 8, 7, 7]

def findThreeLargestNumbers(array):
    # Write your code here.
    threeLargest = [None, None, None]
	for num in array:
		updateLargest(threeLargest, num)
		return threeLargest

# so what we are checing one by one from the largest if these current numbers greater than our current numbers
	
def updateLargest(threeLargest, num):
# 	if threeLargest at index 2 is none, meaning we don't have a number there yet
#	or num our current number is greater than whats in index of 2
#	we want to update that current value 
	if threeLargest[2] is None or num > threeLargest[2]:
		shiftAndUpdate() #Todo
	#then if neither of these conditions are met, meaning we have a number here on the third value
	#then we want to compare to our second biggest number
	elif threeLargest[1] is None or num > threeLargest[1]:
		shiftAndUpdate() #TODO
	elif threeLargest(0) is None or num > threeLargest[0]:
		shiftAndUpdate()#

File: binary-search.js
//  Index  0  1  2    3   4   5  6    7   8   9
//        [0, 1, 21, 33, 45, 45, 61, 71, 72, 73] //looking for number 33
//         L              M                    R
     
            
// M = Middle number or middle pointer
           
// M = Sum of the index L + R / 2 = Average
// M = (0 + 9) / 2 = 4.5
// M = Round down = 4
    
    
// IS 45 Which is the middle number greater than 33? No so element all those numbers

//      [0, 1, 21, 33, 45, 45, 61, 71, 72, 73] //looking for number 33
//       L              M                   R
      

// //Code walkthrough 
// //Soliving in python 
// def binarySearch(array, target):
// return binarySearchHelper(array, target, 0, array - 1)


// def binarySearchHelper(array, target, left, right):



// Given a m x n matrix grid which is sorted in non - increasing order both row - wise and column - wise,
//  return the number of negative numbers in grid.


// Example 1:

// Input: grid = [[4,3,2,-1],[3,2,1,-1],[1,1,-1,-2],[-1,-1,-2,-3]]
// Output: 8
// Explanation: There are 8 negatives number in the matrix.
// Example 2:

// Input: grid = [[3,2],[1,0]]
// Output: 0
// Example 3:

// Input: grid = [[1,-1],[-1,-1]]
// Output: 3